# Detection and Classification of Chronic Kidney Disease

Chronic kidney disease (CKD), also known as chronic renal disease, is progressive loss in kidney function over a period of months or years [1]. If kidney disease gets worse, wastes can build to high levels in your blood and make you feel sick. Patient may develop complications like high blood pressure, anemia (low blood count), weak bones, poor nutritional health and nerve damage. Also, kidney disease increases your risk of having heart and blood vessel disease. These problems may happen slowly over a long period of time.[2]

In this project,Chronic Kidney Disease dataset in UCI Machine learning repository [3] have been explored, which includes 24 attributes excluding the target label class and health parameters of 400 patients

## Knowing the data

In [2]:
import pandas as pd
import numpy as np
from IPython.display import display
from scipy.stats import mode
import matplotlib.pyplot as plt

data = pd.read_csv('ckd.csv',sep=',', na_values=['?'])
# print "Attribute list is {}".format(data.columns.values)

print (" ")
#No of people with chronic kidney disease
n_ckd = len(data[data['class']=='ckd'])

#No of people without chronic kidney disease
n_notckd = len(data[data['class']=='notckd'])



# print "Number of people detected with chronic kidney disease: {}".format(n_ckd)
# print "Number of people not detected with chronic kidney diesease: {}".format(n_notckd)
display(data.describe())

age          bp          sg          al          su         bgr  \
count  391.000000  388.000000  353.000000  354.000000  351.000000  356.000000   
mean    51.483376   76.469072    1.017408    1.016949    0.450142  148.036517   
std     17.169714   13.683637    0.005717    1.352679    1.099191   79.281714   
min      2.000000   50.000000    1.005000    0.000000    0.000000   22.000000   
25%     42.000000   70.000000    1.010000    0.000000    0.000000   99.000000   
50%     55.000000   80.000000    1.020000    0.000000    0.000000  121.000000   
75%     64.500000   80.000000    1.020000    2.000000    0.000000  163.000000   
max     90.000000  180.000000    1.025000    5.000000    5.000000  490.000000   

               bu          sc         sod         pot        hemo         pcv  \
count  381.000000  383.000000  313.000000  312.000000  348.000000  329.000000   
mean    57.425722    3.072454  137.528754    4.627244   12.526437   38.884498   
std     50.503006    5.741126   10.408752    3.193904    2.912587    8.990105   
min      1.500000    0.400000    4.500000    2.500000    3.100000    9.000000   
25%     27.000000    0.900000  135.000000    3.800000   10.300000   32.000000   
50%     42.000000    1.300000  138.000000    4.400000   12.650000   40.000000   
75%     66.000000    2.800000  142.000000    4.900000   15.000000   45.000000   
max    391.000000   76.000000  163.000000   47.000000   17.800000   54.000000   

               wbcc        rbcc  
count    294.000000  269.000000  
mean    8406.122449    4.707435  
std     2944.474190    1.025323  
min     2200.000000    2.100000  
25%     6500.000000    3.900000  
50%     8000.000000    4.800000  
75%     9800.000000    5.400000  
max    26400.000000    8.000000

In [3]:
from pandas.tools.plotting import scatter_matrix
scatter_matrix(data, alpha=0.6, figsize=(12, 12), diagonal='kde')
plt.show()

import seaborn as sns
plt.figure(figsize=(30,20))
sns.heatmap(data.corr(), annot=True)
plt.show()

ModuleNotFoundError: No module named 'pandas.tools'

We have 24 feature attributes with 11 numerical attributes and 14 nominal and one target label.The number of people detected with chronic kidney disease(ckd) is 250 and with no-ckd is 150. In the scatter plot, hemo,pcv and rbcc shows high positive correlation.

Handling the Missing values

In [4]:
#Filling missing value with most frequent for nominal and median for numerical

X = pd.DataFrame(data)
fill = pd.Series([X[c].value_counts().index[0]
        if X[c].dtype == np.dtype('O') else X[c].median() for c in X],
        index=X.columns)
new_data=X.fillna(fill) 
print (new_data )           

      age    bp     sg   al   su     rbc        pc         pcc          ba  \
0    48.0  80.0  1.020  1.0  0.0  normal    normal  notpresent  notpresent   
1     7.0  50.0  1.020  4.0  0.0  normal    normal  notpresent  notpresent   
2    62.0  80.0  1.010  2.0  3.0  normal    normal  notpresent  notpresent   
3    48.0  70.0  1.005  4.0  0.0  normal  abnormal     present  notpresent   
4    51.0  80.0  1.010  2.0  0.0  normal    normal  notpresent  notpresent   
..    ...   ...    ...  ...  ...     ...       ...         ...         ...   
395  55.0  80.0  1.020  0.0  0.0  normal    normal  notpresent  notpresent   
396  42.0  70.0  1.025  0.0  0.0  normal    normal  notpresent  notpresent   
397  12.0  80.0  1.020  0.0  0.0  normal    normal  notpresent  notpresent   
398  17.0  60.0  1.025  0.0  0.0  normal    normal  notpresent  notpresent   
399  58.0  80.0  1.025  0.0  0.0  normal    normal  notpresent  notpresent   

       bgr  ...   pcv    wbcc  rbcc  htn   dm  cad  appet   pe 

Label Encoding on the nominal values

In [ ]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
data=new_data.copy()
for items in data:
    if data[items].dtype == np.dtype('O'):
        data[items]=le.fit_transform(data[items])

print data.dtypes
print data

age      float64
bp       float64
sg       float64
al       float64
su       float64
rbc        int64
pc         int64
pcc        int64
ba         int64
bgr      float64
bu       float64
sc       float64
sod      float64
pot      float64
hemo     float64
pcv      float64
wbcc     float64
rbcc     float64
htn        int64
dm         int64
cad        int64
appet      int64
pe         int64
ane        int64
class      int64
dtype: object
      age     bp     sg   al   su  rbc  pc  pcc  ba    bgr  ...     pcv  \
0    48.0   80.0  1.020  1.0  0.0    1   1    0   0  121.0  ...    44.0   
1     7.0   50.0  1.020  4.0  0.0    1   1    0   0  121.0  ...    38.0   
2    62.0   80.0  1.010  2.0  3.0    1   1    0   0  423.0  ...    31.0   
3    48.0   70.0  1.005  4.0  0.0    1   0    1   0  117.0  ...    32.0   
4    51.0   80.0  1.010  2.0  0.0    1   1    0   0  106.0  ...    35.0   
5    60.0   90.0  1.015  3.0  0.0    1   1    0   0   74.0  ...    39.0   
6    68.0   70.0  1.010  0.0  0.0   

Scaling features using Robust Scaler 

In [ ]:
from sklearn.preprocessing import StandardScaler, RobustScaler
target_class = data['class']
features = data.drop('class', axis = 1)
data_robust = pd.DataFrame(RobustScaler().fit_transform(features), columns=features.columns)


Dimensionality Reduction

In [ ]:
from sklearn.decomposition import PCA
pca = PCA()
data_pca = pd.DataFrame(pca.fit_transform(data_robust), columns=data_robust.columns)
print data_robust.shape
print pca.explained_variance_ratio_


(400, 24)
[ 0.38157857  0.25203875  0.10337358  0.05525873  0.04128847  0.0322456
  0.02664966  0.02054982  0.01970661  0.01407955  0.01267326  0.00877071
  0.00745055  0.00447937  0.00329693  0.00259371  0.00241495  0.00227722
  0.00216587  0.00189897  0.00181543  0.00136625  0.0010538   0.00097364]


First 11 components gives principal components that explain at least 95% of total variance. So let the number of components be 11.

In [ ]:
pca = PCA(n_components=11)
pca.fit(data_robust)
reduced_data = pca.transform(data_robust)
reduced_data = pd.DataFrame(reduced_data, columns = ['dim1','dim2','dim3','dim4','dim5','dim6','dim7','dim8','dim9','dim10','dim11'])
print reduced_data.head()

       dim1      dim2      dim3      dim4      dim5      dim6      dim7  \
0 -1.127701  0.779780  0.642923  0.190267  0.609335  0.444339  0.020288   
1 -1.408005  1.026680  1.063298 -0.043320 -2.451236  0.824565 -0.597719   
2  0.248468 -0.277980 -5.114364  3.851732 -0.211125 -0.270017 -0.949867   
3 -0.491470 -3.934303 -1.223619 -1.503578 -1.921622  3.695031  0.158452   
4 -0.775043  0.395504 -0.290283 -0.980759  0.116809  0.880136  0.586873   

       dim8      dim9     dim10     dim11  
0 -0.280839 -0.007856 -0.222121 -0.000352  
1  0.786148  0.299144  1.906853  0.017050  
2  0.489734  0.624682  0.248006 -0.615226  
3 -0.023544  1.851161 -0.102150  0.294735  
4  1.217560  1.012095 -0.188106 -0.184063  


Splitting the dataset to training and testing set

In [ ]:

from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(reduced_data,target_class, test_size=0.25, random_state=42)



C:\Users\SREE\Anaconda2\lib\site-packages\sklearn\cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


The Six Classifiers without hypertuning parameters

In [ ]:
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer
from sklearn.metrics import f1_score
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.cross_validation import StratifiedShuffleSplit
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn import metrics
from sklearn import tree
from sklearn.neural_network import MLPClassifier
from sklearn.naive_bayes import GaussianNB
from time import time
clf_A=GaussianNB()
clf_B=tree.DecisionTreeClassifier(random_state=0)
clf_C=SVC()
clf_D=RandomForestClassifier()
clf_E=KNeighborsClassifier()
clf_F=MLPClassifier(random_state=1)

for clf in [clf_A, clf_B, clf_C, clf_D, clf_E, clf_F]:
    
    print "\n{}: \n".format(clf.__class__.__name__)
    start = time()
  
    clf.fit(X_train, y_train)
    
    end = time()
    print "Trained model in {:.4f} milliseconds".format((end - start)*1000)
    
for clf in [clf_A, clf_B, clf_C, clf_D, clf_E, clf_F]:
    y_pred = clf.predict(X_test)
    f1_scorer=f1_score(y_test, y_pred, pos_label=0)
    print "\n{}: \n".format(clf.__class__.__name__)
    print "f1 score for test test is {}".format(f1_scorer)


C:\Users\SREE\Anaconda2\lib\site-packages\sklearn\grid_search.py:43: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)



GaussianNB: 

Trained model in 2.0001 milliseconds

DecisionTreeClassifier: 

Trained model in 1.9999 milliseconds

SVC: 

Trained model in 2.0001 milliseconds

RandomForestClassifier: 

Trained model in 27.9999 milliseconds

KNeighborsClassifier: 

Trained model in 0.0000 milliseconds

MLPClassifier: 

Trained model in 374.0001 milliseconds

GaussianNB: 

f1 score for test test is 0.96

DecisionTreeClassifier: 

f1 score for test test is 0.954545454545

SVC: 

f1 score for test test is 0.984375

RandomForestClassifier: 

f1 score for test test is 0.969696969697

KNeighborsClassifier: 

f1 score for test test is 0.992248062016

MLPClassifier: 

f1 score for test test is 0.992248062016


C:\Users\SREE\Anaconda2\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:563: ConvergenceWarning: Stochastic Optimizer: Maximum iterations reached and the optimization hasn't converged yet.
  % (), ConvergenceWarning)


| Classifier        | Training Time (ms)         | F1 Score  |
| ------------------|:--------------------------:| ---------:|
| Gaussian NB       | 0.0000                     | 0.96      |
| Decision Tree     | 0.0000                     | 0.9545    |
| SVC               | 15.0001                    | 0.9843    |
| Random Forest     | 46.9999                    | 0.9767    |
| KNN               | 0.000                      | 0.9922    |
| MLP               | 375.9999                   | 0.9922    |

Classifiers with hypertuning

In [ ]:
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer
from sklearn.metrics import f1_score
from sklearn.svm import SVC
from sklearn.cross_validation import StratifiedShuffleSplit
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn import metrics


parameters = [{'kernel':['rbf','sigmoid'],'C': [.1,1,10], 'gamma': [0.001,10,1000]},]
clf = SVC()
f1_scorer = make_scorer(f1_score,pos_label=0)
sss = StratifiedShuffleSplit( y_train, n_iter=10, test_size=0.25)
grid_obj = GridSearchCV(clf,parameters,cv = sss,scoring=f1_scorer)
grid_obj = grid_obj.fit(X_train,y_train)
clf = grid_obj.best_estimator_
#print clf
y_pred = clf.predict(X_test)
f1_score_value=f1_score(y_test, y_pred, pos_label=0) # For testing
print "F1 Score for test set: {}".format(f1_score_value)
print "Confusion Matrix is : \n  {} ".format(confusion_matrix(y_test, y_pred))
target_names = ['class 0', 'class 1']
print " "
print "Classification report is : \n  "
print(classification_report(y_test, y_pred, target_names=target_names))

        

F1 Score for test set: 0.984375
Confusion Matrix is : 
  [[63  2]
 [ 0 35]] 
 
Classification report is : 
  
             precision    recall  f1-score   support

    class 0       1.00      0.97      0.98        65
    class 1       0.95      1.00      0.97        35

avg / total       0.98      0.98      0.98       100



Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
parameters = [{'n_estimators': [10,50,75],
               'max_features': ["auto","sqrt","log2"],'n_jobs': [1]}]
clf = RandomForestClassifier()
f1_scorer = make_scorer(f1_score,pos_label=0)
sss = StratifiedShuffleSplit( y_train, n_iter=10, test_size=0.25)
grid_obj = GridSearchCV(clf,parameters,cv = sss,scoring=f1_scorer)
grid_obj = grid_obj.fit(X_train,y_train)
clf = grid_obj.best_estimator_
y_pred = clf.predict(X_test)
f1_score_value=f1_score(y_test, y_pred, pos_label=0) # For testing
print "F1 Score for test set: {}".format(f1_score_value)
print "Confusion Matrix is : \n  {} ".format(confusion_matrix(y_test, y_pred))
target_names = ['class 0', 'class 1']
print " "
print "Classification report is : \n  "
print(classification_report(y_test, y_pred, target_names=target_names))

F1 Score for test set: 0.969230769231
Confusion Matrix is : 
  [[63  2]
 [ 2 33]] 
 
Classification report is : 
  
             precision    recall  f1-score   support

    class 0       0.97      0.97      0.97        65
    class 1       0.94      0.94      0.94        35

avg / total       0.96      0.96      0.96       100



KNeighborsClassifier

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
parameters = [{'metric':['minkowski','euclidean'] ,
               'weights':['uniform'],'n_neighbors': [5,15,10]}]
clf = KNeighborsClassifier()
f1_scorer = make_scorer(f1_score,pos_label=0)
sss = StratifiedShuffleSplit( y_train, n_iter=10, test_size=0.25)
grid_obj = GridSearchCV(clf,parameters,cv = sss,scoring=f1_scorer)
grid_obj = grid_obj.fit(X_train,y_train)
clf = grid_obj.best_estimator_
print clf
y_pred = clf.predict(X_test)
f1_score_value=f1_score(y_test, y_pred, pos_label=0) # For testing
print "F1 Score for test set: {}".format(f1_score_value)
print "Confusion Matrix is : \n  {} ".format(confusion_matrix(y_test, y_pred))
target_names = ['class 0', 'class 1']
print " "
print "Classification report is : \n  "
print(classification_report(y_test, y_pred, target_names=target_names))

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')
F1 Score for test set: 0.992248062016
Confusion Matrix is : 
  [[64  1]
 [ 0 35]] 
 
Classification report is : 
  
             precision    recall  f1-score   support

    class 0       1.00      0.98      0.99        65
    class 1       0.97      1.00      0.99        35

avg / total       0.99      0.99      0.99       100



Decision Tree

In [ ]:
from sklearn import tree
from sklearn.metrics import classification_report
parameters = [{'criterion':['entropy','gini'] ,'splitter':['best','random']}]
clf = tree.DecisionTreeClassifier(random_state=0)
f1_scorer = make_scorer(f1_score,pos_label=0)
sss = StratifiedShuffleSplit( y_train, n_iter=10, test_size=0.25)
grid_obj = GridSearchCV(clf,parameters,cv = sss,scoring=f1_scorer)
grid_obj = grid_obj.fit(X_train,y_train)
clf = grid_obj.best_estimator_
y_pred = clf.predict(X_test)
f1_score_value=f1_score(y_test, y_pred, pos_label=0) # For testing
print "F1 Score for test set: {}".format(f1_score_value)
print "Confusion Matrix is : \n  {} ".format(confusion_matrix(y_test, y_pred))
target_names = ['class 0', 'class 1']
print " "
print "Classification report is : \n  "
print(classification_report(y_test, y_pred, target_names=target_names))

F1 Score for test set: 0.992248062016
Confusion Matrix is : 
  [[64  1]
 [ 0 35]] 
 
Classification report is : 
  
             precision    recall  f1-score   support

    class 0       1.00      0.98      0.99        65
    class 1       0.97      1.00      0.99        35

avg / total       0.99      0.99      0.99       100



In [ ]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import make_scorer
from sklearn.metrics import f1_score
from sklearn.grid_search import GridSearchCV
from sklearn.cross_validation import StratifiedShuffleSplit
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn import metrics
parameters = [{'solver':['lbfgs','sgd'],'alpha':[1e-6,1e-3],'hidden_layer_sizes':[3,6], }]
clf = MLPClassifier(random_state=42)
f1_scorer = make_scorer(f1_score,pos_label=0)
sss = StratifiedShuffleSplit( y_train, n_iter=10, test_size=0.25)
grid_obj = GridSearchCV(clf,parameters,cv = sss,scoring=f1_scorer)
grid_obj = grid_obj.fit(X_train,y_train)
clf = grid_obj.best_estimator_
y_pred = clf.predict(X_test)
f1_score_value=f1_score(y_test, y_pred, pos_label=0) # For testing
print "F1 Score for test set: {}".format(f1_score_value)
print "Confusion Matrix is : \n  {} ".format(confusion_matrix(y_test, y_pred))
target_names = ['class 0', 'class 1']
print " "
print "Classification report is : \n  "
print(classification_report(y_test, y_pred, target_names=target_names))



F1 Score for test set: 0.992248062016
Confusion Matrix is : 
  [[64  1]
 [ 0 35]] 
 
Classification report is : 
  
             precision    recall  f1-score   support

    class 0       1.00      0.98      0.99        65
    class 1       0.97      1.00      0.99        35

avg / total       0.99      0.99      0.99       100



In [ ]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
y_pred = gnb.fit(X_train,y_train).predict(X_test)

f1_score_value=f1_score(y_test, y_pred, pos_label=0) # For testing
print "F1 Score for test set: {}".format(f1_score_value)
print "Confusion Matrix is : \n  {} ".format(confusion_matrix(y_test, y_pred))
target_names = ['class 0', 'class 1']
print " "
print "Classification report is : \n  "
print(classification_report(y_test, y_pred, target_names=target_names))


F1 Score for test set: 0.96
Confusion Matrix is : 
  [[60  5]
 [ 0 35]] 
 
Classification report is : 
  
             precision    recall  f1-score   support

    class 0       1.00      0.92      0.96        65
    class 1       0.88      1.00      0.93        35

avg / total       0.96      0.95      0.95       100



With Naive Bayes(NB) as a benchmark of evaluation with F1 score of NB being 0.96, KNN outperforms NB with 0.99 with and without hypertuning and completing the training in less time.ANN also outperforms NB in terms of F1 score but it takes much training time even without hypertuning compared to others.Decision Tree shows an increase in performance with parameter optimization.
So comparing to the benchmark NB, all classifiers performs well but KNN performs really well with better accuracy and less time as medical diagnostic systems must give high performance and accuracy.



References:
    
    [1] https://en.wikipedia.org/wiki/Chronic_kidney_disease
    [2] https://www.kidney.org/kidneydisease/aboutckd
    [3]  L. Rubini, “Early stage of chronic kidney disease UCI machine learning repository,” 2015. [Online]. Available:           http://archive.ics.uci.edu/ml/datasets/Chronic Kidney Disease 
    [4] http://www.worldkidneyday.org/faqs/chronic-kidney-disease/
    [5] http://scikit-learn.org/stable/modules/model_evaluation.html
    [6] K.M. Al-Aidaroos, A.A. Bakar and Z. Othman ,Medical Data Classification with Naive Bayes Approach

   